File: 01-ordenacao/csharp/OrdenacaoMVC/src/View/MenuView.cs
using OrdenacaoMVC.Controller;
using OrdenacaoMVC.Model;

namespace OrdenacaoMVC.View;

/// <summary>
/// Exibe o menu e interage com o usuário via console
/// Toda a leitura/escrita do programa passa por aqui
///</summary>
public class MenuView
{
    private const int OpcaoGerar = 1;
    private const int OpcaoMostrar = 2;
    private const int OpcaoCompararTodos = 3;
    private const int PrimeiraOpcaoAlgoritmo = 4;

    private readonly OrdenacaoController controller;

    public MenuView(OrdenacaoController controller)
    {
        this.controller = controller;
    }

    public void Iniciar()
    {
        int opcao;

        do
        {
            ExibirMenu();

            string? entrada = Console.ReadLine();

            if (!int.TryParse(entrada, out opcao))
            {
                Console.WriteLine("Entrada inválida! Digite um número.");
                continue;
            }

            if (opcao == OpcaoGerar)
            {
                GerarLista();
            }
            else if (opcao == OpcaoMostrar)
            {
                MostrarLista();
            }
            else if (opcao == OpcaoCompararTodos)
            {
                CompararTodos();
            }
            else if (opcao >= PrimeiraOpcaoAlgoritmo &&
                     opcao < PrimeiraOpcaoAlgoritmo + controller.Algoritmos.Count)
            {
                ExecutarAlgoritmo(opcao - PrimeiraOpcaoAlgoritmo + 1);
            }
            else if (opcao == OpcaoSair())
            {
                Console.WriteLine("Encerrando...");
            }
            else
            {
                Console.WriteLine("Opção inválida!");
            }

        } while (opcao != OpcaoSair());
    }

    private int OpcaoSair() => PrimeiraOpcaoAlgoritmo + controller.Algoritmos.Count;

    private void ExibirMenu()
    {
        Console.WriteLine("--- MENU ---");
        Console.WriteLine($"{OpcaoGerar} - Gerar nova lista");
        Console.WriteLine($"{OpcaoMostrar} - Mostrar lista atual");
        Console.WriteLine($"{OpcaoCompararTodos} - Comparar todos os algoritmos (cópia independente)");
        Console.WriteLine($"{PrimeiraOpcaoAlgoritmo} - Ordenar por {controller.Algoritmos[0].Nome}");

        for (int i = 1; i < controller.Algoritmos.Count; i++)
        {
            Console.WriteLine($"{PrimeiraOpcaoAlgoritmo + i} - Ordenar por {controller.Algoritmos[i].Nome}");
        }

        Console.WriteLine($"{OpcaoSair()} - Sair");
        Console.Write("Escolha uma opção: ");
    }

    private void GerarLista()
    {
        Console.Write("Quantos números deseja gerar? ");
        string? entrada = Console.ReadLine();

        if (!int.TryParse(entrada, out int quantidade) || quantidade <= 0)
        {
            Console.WriteLine("Quantidade inválida!");
            return;
        }

        controller.GerarLista(quantidade, 0, 100000);
        Console.WriteLine($"Lista gerada com {quantidade} números.");
    }

    private void MostrarLista()
    {
        IReadOnlyList<int> lista = controller.ObterLista();

        if (lista.Count == 0)
        {
            Console.WriteLine("\nNenhuma lista carregada. Gere uma primeiro.");
            return;
        }

        Console.WriteLine($"\nLista atual ({lista.Count} itens):");

        foreach (int item in lista)
        {
            Console.WriteLine(item);
        }
    }

    private void ExecutarAlgoritmo(int indice1Based)
    {
        ResultadoOrdenacao? resultado = controller.ExecutarAlgoritmo(indice1Based);

        if (resultado is null)
        {
            Console.WriteLine("Falha ao executar o algoritmo.");
            return;
        }

        string nomeAlgoritmo = controller.Algoritmos[indice1Based - 1].Nome;

        Console.WriteLine($"\n{nomeAlgoritmo}:");
        Console.WriteLine($"  Comparações:   {resultado.Comparacoes}");
        Console.WriteLine($"  Trocas:        {resultado.Trocas}");
        Console.WriteLine($"  Movimentações: {resultado.Movimentacoes}");
        Console.WriteLine($"  Tempo (ms):    {resultado.TempoMs}");
        Console.WriteLine($"  Tamanho:       {resultado.TamanhoLista}");
    }

    private void CompararTodos()
    {
        var resultados = controller.ExecutarTodosAlgoritmos();

        if (resultados.Count == 0)
        {
            Console.WriteLine("\nNenhuma lista carregada. Gere uma primeiro.");
            return;
        }

        Console.WriteLine("\n=== COMPARAÇÃO EXPERIMENTAL (cópia independente) ===");
        Console.WriteLine("Algoritmo                    Comparações   Trocas  Movimentações  Tempo (ms)");
        Console.WriteLine("--------------------------------------------------------------------------------");

        var algoritmos = controller.Algoritmos;
        for (int i = 0; i < algoritmos.Count; i++)
        {
            string nome = algoritmos[i].Nome;
            var r = resultados[i];

            Console.WriteLine($"{nome,-30} {r.Comparacoes,12} {r.Trocas,7} {r.Movimentacoes,13} {r.TempoMs,10}");
        }
    }
}

File: 01-ordenacao/java/OrdenacaoMVC/src/view/MenuView.java
package view;

import controller.OrdenacaoController;
import java.util.List;
import java.util.Optional;
import java.util.Scanner;
import model.ResultadoOrdenacao;
import model.algoritmos.IAlgoritmoOrdenacao;

/**
 * Exibe o menu e interage com o usuário via console
 */
public class MenuView {

    private static final int OPCAO_GERAR = 1;
    private static final int OPCAO_MOSTRAR = 2;
    private static final int OPCAO_COMPARAR_TODOS = 3;
    private static final int PRIMEIRA_OPCAO_ALGORITMO = 4;

    private final OrdenacaoController controller;
    private final Scanner scanner;

    public MenuView(OrdenacaoController controller) {
        this.controller = controller;
        this.scanner = new Scanner(System.in);
    }

    public void iniciar() {
        int opcao = 0;

        do {
            exibirMenu();

            String entrada = scanner.nextLine();

            try {
                opcao = Integer.parseInt(entrada);
            } catch (NumberFormatException e) {
                System.out.println("Entrada inválida! Digite um número.");
                continue;
            }

            if (opcao == OPCAO_GERAR) {
                gerarLista();
            } else if (opcao == OPCAO_MOSTRAR) {
                mostrarLista();
            } else if (opcao == OPCAO_COMPARAR_TODOS) {
                compararTodos();
            } else if (opcao >= PRIMEIRA_OPCAO_ALGORITMO &&
                       opcao < PRIMEIRA_OPCAO_ALGORITMO + controller.getAlgoritmos().size()) {
                executarAlgoritmo(opcao - PRIMEIRA_OPCAO_ALGORITMO + 1); // converte nº do menu para índice 1-based
            } else if (opcao == opcaoSair()) {
                System.out.println("Encerrando...");
            } else {
                System.out.println("Opção inválida!");
            }

        } while (opcao != opcaoSair());
    }

    private int opcaoSair() {
        return PRIMEIRA_OPCAO_ALGORITMO + controller.getAlgoritmos().size();
    }

    private void exibirMenu() {
        System.out.println("--- MENU ---");
        System.out.println(OPCAO_GERAR + " - Gerar nova lista");
        System.out.println(OPCAO_MOSTRAR + " - Mostrar lista atual");
        System.out.println(OPCAO_COMPARAR_TODOS + " - Comparar todos os algoritmos (cópia independente)");
        System.out.println(PRIMEIRA_OPCAO_ALGORITMO + " - Ordenar por " + controller.getAlgoritmos().get(0).getNome());

        for (int i = 1; i < controller.getAlgoritmos().size(); i++) {
            System.out.println(PRIMEIRA_OPCAO_ALGORITMO + i + " - Ordenar por " + controller.getAlgoritmos().get(i).getNome());
        }

        System.out.println(opcaoSair() + " - Sair");
        System.out.print("Escolha uma opção: ");
    }

    private void gerarLista() {
        System.out.print("Quantos números deseja gerar? ");
        String entrada = scanner.nextLine();

        int quantidade;
        try {
            quantidade = Integer.parseInt(entrada);
        } catch (NumberFormatException e) {
            System.out.println("Quantidade inválida!");
            return;
        }

        if (quantidade <= 0) {
            System.out.println("Quantidade inválida!");
            return;
        }

        controller.gerarLista(quantidade, 0, 100000);
        System.out.println("Lista gerada com " + quantidade + " números.");
    }

    private void mostrarLista() {
        List<Integer> lista = controller.getLista();

        if (lista.isEmpty()) {
            System.out.println("\nNenhuma lista carregada. Gere uma primeiro.");
            return;
        }

        System.out.println("\nLista atual (" + lista.size() + " itens):");

        for (int item : lista) {
            System.out.println(item);
        }
    }

    /**
     * Método que executa o algoritmo de ordenação pelo índice
     * @param indice1Based índice do algoritmo (1-based)
     */
    private void executarAlgoritmo(int indice1Based) {
        Optional<ResultadoOrdenacao> resultado = controller.executarAlgoritmo(indice1Based);

        if (resultado.isEmpty()) {
            System.out.println("Falha ao executar o algoritmo.");
            return;
        }

        ResultadoOrdenacao r = resultado.get();
        String nomeAlgoritmo = controller.getAlgoritmos().get(indice1Based - 1).getNome();

        System.out.println("\n" + nomeAlgoritmo + ":");
        System.out.println("  Comparações:   " + r.comparacoes());
        System.out.println("  Trocas:        " + r.trocas());
        System.out.println("  Movimentações: " + r.movimentacoes());
        System.out.println("  Tempo (ms):    " + r.tempoMs());
        System.out.println("  Tamanho:       " + r.tamanhoLista());
    }

    private void compararTodos() {
        List<ResultadoOrdenacao> resultados = controller.executarTodosAlgoritmos();

        if (resultados.isEmpty()) {
            System.out.println("\nNenhuma lista carregada. Gere uma primeiro.");
            return;
        }

        System.out.println("\n=== COMPARAÇÃO EXPERIMENTAL (cópia independente) ===");
        System.out.println("Algoritmo                    Comparações   Trocas  Movimentações  Tempo (ms)");
        System.out.println("--------------------------------------------------------------------------------");

        List<IAlgoritmoOrdenacao> algoritmos = controller.getAlgoritmos();
        for (int i = 0; i < algoritmos.size(); i++) {
            String nome = algoritmos.get(i).getNome();
            ResultadoOrdenacao r = resultados.get(i);

            System.out.printf("%-30s %12d %7d %13d %10d%n",
                nome, r.comparacoes(), r.trocas(), r.movimentacoes(), r.tempoMs());
        }
    }
}
